code/modules/xenoarchaeology/tools/console/xenoarchaeology_console.dm

```diff
@@ -0,0 +1,261 @@
+			if(length(console_orders))
+				current_order = console_orders[length(console_orders)]
+				current_pack = console_orders[current_order]
+				if(length(current_pack.contains) < max_pack_contents && (current_order in SSsupply.shoppinglist))
+					//Update pack content and cost
+					current_pack.cost += seller?.get_price(locate(params["item_id"]))
+					current_pack.contains += seller?.buy_stock(locate(params["item_id"]))
+					//Generate a new order
+					SSsupply.shoppinglist -= current_order
+					console_orders -= current_order
+					qdel(current_order)
+					current_order = new /datum/supply_order(current_pack, name, rank, ckey, "Research Material Requisition", D)
+					SSsupply.shoppinglist += current_order
+					console_orders[current_order] = current_pack
+					ui_update()
+					return
+				else if(!(current_order in SSsupply.shoppinglist))
+					console_orders -= current_order
+					qdel(current_order)
+			//If we can't, make a new order
+			current_pack = new /datum/supply_pack/science_listing()
+			current_pack.contains = list()
+			current_pack?.current_supply = max(1, current_pack.current_supply) //Don't worry about it :^)
+			current_pack.cost += seller?.get_price(locate(params["item_id"]))
+			current_pack.contains += seller?.buy_stock(locate(params["item_id"]))
+			current_order = new /datum/supply_order(current_pack, name, rank, ckey, "Research Material Requisition", D)
+			current_order.generateRequisition(get_turf(src))
+			SSsupply.shoppinglist += current_order
+			console_orders[current_order] = current_pack
+		//Radio shit
+		if("toggle_purchase_audio")
+			radio_purchase_notice = !radio_purchase_notice
+		if("toggle_solved_audio")
+			radio_solved_notice = !radio_solved_notice
+
+	ui_update()
+
+/obj/machinery/computer/xenoarchaeology_console/proc/check_sold(datum/source, atom/movable/AM, sold)
+	SIGNAL_HANDLER
+
+	var/obj/item/sticker/xenoartifact_label/L = AM
+	if(!istype(L))
+		return
+	var/atom/artifact = L.loc
+	var/datum/component/xenoartifact/X = artifact?.GetComponent(/datum/component/xenoartifact)
+	if(!X || !artifact)
+		return
+	//Grab values to calculate success
+	var/score = 0
+	var/max_score = 0
+	var/bonus = 0
+	var/max_bonus = 0
+	var/attempted_bonus = FALSE
+	var/list/traits_by_type = list()
+	for(var/i in X.artifact_traits) //By priority
+		for(var/datum/xenoartifact_trait/T in X.artifact_traits[i]) //By trait in priorty
+			traits_by_type += list(T.type)
+			if(T.contribute_calibration)
+				max_score += 1
+			else
+				max_bonus += 1
+	for(var/datum/xenoartifact_trait/T as() in L.traits)
+		if((T in traits_by_type))
+			if(initial(T.contribute_calibration))
+				score += 1
+			else
+				bonus += 1
+				attempted_bonus = TRUE
+		else
+			if(initial(T.contribute_calibration))
+				score -= 1
+			else
+				bonus -= 1
+				attempted_bonus = TRUE
+	//Calculate success rate
+	var/success_rate = score / (max_score || 1)
+	var/bonus_rate = max(1, 2*(bonus/(max_bonus||1)))
```

Comment:
before this, `max_bonus` and `max_score` should do `max_thing = max(max_thing, 1)`, rather than trying `max_thing || 1`.